File: python/app/vulns/access_control.py
"""Broken Access Control — OWASP 2021 A01  ->  OWASP 2025 A01.

Access-control flaws are *authorization* bugs: the request is understood, often
even authenticated, but the server fails to check whether *this* caller may act
on *this* object or reach *this* function. Unlike injection, no framework or ORM
fixes them for you — object- and function-level authorization is application
logic that must be written (and reviewed) by hand.

The comment/annotation format here mirrors the reference module
``app/vulns/injection_sql.py``: each permutation carries a standard header with
its CWE, an exploitation-likelihood rating, prevalence notes, an example
request, and the corresponding safe pattern.
"""
from flask import Blueprint, request, jsonify

from app.db import get_db

bp = Blueprint("access_control", __name__, url_prefix="/access")


# =============================================================================
# PERMUTATION 1 — IDOR: read any note by id, no ownership check
# OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
# CWE-639: Authorization Bypass Through User-Controlled Key
# EXPLOITATION LIKELIHOOD: HIGH — a single integer parameter selects the object;
#   ids are sequential so an attacker just enumerates 1,2,3,… to walk every
#   record, private ones included. No auth, no tooling, fully deterministic.
# PREVALENCE TODAY: greenfield HIGH (ORMs/frameworks are safe-by-default against
#   SQLi but provide NO object-level ownership check — `Note.get(id)` happily
#   returns anyone's row) | legacy/10yr tech-debt HIGH (raw id->row lookups with
#   authorization "handled elsewhere" are endemic).
# EXAMPLE:
#   GET /access/note?id=2   -> alice's private "Bank PIN reminder"
#   GET /access/note?id=4   -> admin's private "Ops runbook" (prod db password)
# FIX: scope every lookup to the caller — WHERE id=? AND (private=0 OR owner=?)
#      — and verify ownership before returning (see /access/note-safe).
# =============================================================================
@bp.get("/note")
def note():
    note_id = request.args.get("id", "")
    db = get_db()
    # VULNERABLE: the note is fetched and returned for ANY id with no check that
    # the caller owns it or that it is public — the private flag is ignored.
    row = db.execute(
        "SELECT id, owner, title, body, private FROM notes WHERE id = ?",
        (note_id,),
    ).fetchone()
    if row is None:
        return jsonify({"id": note_id, "note": None})
    return jsonify({"id": note_id, "note": dict(row)})


# =============================================================================
# PERMUTATION 2 — Missing function-level authorization on an admin endpoint
# OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
# CWE-862: Missing Authorization
# EXPLOITATION LIKELIHOOD: HIGH — the "admin" URL has zero authentication or
#   role gate; anyone who knows (or guesses) the path dumps every user's md5
#   password hash, email and SSN. Pre-auth, one request, catastrophic impact.
# PREVALENCE TODAY: greenfield MEDIUM (auth middleware/decorators are standard,
#   but newly-added admin/internal/debug routes routinely ship without the guard
#   — "protected by being unlinked") | legacy/10yr tech-debt HIGH (authorization
#   scattered across handlers by hand, trivially forgotten on new endpoints).
# EXAMPLE:
#   curl http://127.0.0.1:5000/access/admin/users
#     -> [{username:admin, password:<md5>, ssn:999-99-9999, role:admin}, …]
# FIX: require authentication AND an explicit role check (deny-by-default) on
#      every privileged route, enforced by middleware — never by obscurity.
# =============================================================================
@bp.get("/admin/users")
def admin_users():
    db = get_db()
    # VULNERABLE: privileged data returned with no authentication or role check.
    rows = [
        dict(r)
        for r in db.execute(
            "SELECT id, username, password, email, role, ssn FROM users"
        ).fetchall()
    ]
    return jsonify({"note": "no auth/role check performed", "users": rows})


# =============================================================================
# PERMUTATION 3 — Mass assignment / privilege escalation
# OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
# CWE-915: Improperly Controlled Modification of Dynamically-Determined Object
#          Attributes (Mass Assignment / autobinding)
# EXPLOITATION LIKELIHOOD: HIGH — the handler writes EVERY field in the request
#   body to the user's row, so a normal user smuggles an extra `role` key and
#   promotes themselves to admin. One authenticated POST, no special tooling.
# PREVALENCE TODAY: greenfield MEDIUM (explicit DTOs/serializers with an allow-
#   list of fields help, but "spread req.body into the model" and frameworks
#   that auto-bind request params to entities keep it alive) | legacy/10yr
#   tech-debt HIGH (whole-object updates from the form/body are the default).
# EXAMPLE:
#   curl -X POST -H 'Content-Type: application/json' \
#     -d '{"username":"alice","email":"a@x.com","role":"admin"}' \
#     http://127.0.0.1:5000/access/profile      -> alice.role becomes "admin"
# FIX: bind only a fixed allow-list of user-editable columns (username, email);
#      never let a privileged column like `role` be set from the request body.
# =============================================================================
@bp.post("/profile")
def profile():
    data = request.get_json(silent=True)
    if data is None:
        data = request.form.to_dict()
    data = data or {}
    username = data.get("username", "")
    # VULNERABLE: every key present in the body (except the selector) is written
    # straight to the row — including privileged columns such as `role`.
    updates = {col: val for col, val in data.items() if col != "username"}
    set_clause = ", ".join(f"{col} = ?" for col in updates)
    sql = f"UPDATE users SET {set_clause} WHERE username = ?"
    params = list(updates.values()) + [username]
    db = get_db()
    try:
        db.execute(sql, params)
        db.commit()
    except Exception as exc:
        return jsonify({"query": sql, "error": str(exc)}), 500
    row = db.execute(
        "SELECT id, username, email, role, ssn FROM users WHERE username = ?",
        (username,),
    ).fetchone()
    return jsonify({"query": sql, "updated": dict(row) if row else None})


# =============================================================================
# PERMUTATION 4 — Client-controlled authorization (trusting a client value)
# OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
# CWE-602: Client-Side Enforcement of Server-Side Security
# EXPLOITATION LIKELIHOOD: HIGH — the admin decision is read from an attacker-
#   supplied `X-Role` header (or `?role=` param) and trusted verbatim; sending
#   `X-Role: admin` unlocks admin-only data. Trivial, deterministic, pre-auth.
# PREVALENCE TODAY: greenfield LOW (server-verified session/JWT claims are the
#   safe default; a raw header deciding a role is an obvious review smell) —
#   though microservice "trust the gateway header" designs reintroduce it |
#   legacy/10yr tech-debt HIGH (hand-rolled auth reading a role from a header,
#   cookie value, or hidden form field the client can just rewrite).
# EXAMPLE:
#   curl -H 'X-Role: admin' http://127.0.0.1:5000/access/dashboard
#   GET /access/dashboard?role=admin
# FIX: derive the role from a server-side, cryptographically-verified session or
#      signed token — never from a header/param/cookie the client can set.
# =============================================================================
@bp.get("/dashboard")
def dashboard():
    # VULNERABLE: authorization decided from a value the client fully controls.
    role = request.headers.get("X-Role") or request.args.get("role", "user")
    if role == "admin":
        db = get_db()
        secret = [
            dict(r)
            for r in db.execute(
                "SELECT username, password, ssn FROM users"
            ).fetchall()
        ]
        return jsonify({"role": role, "admin": True, "secret_data": secret})
    return jsonify({"role": role, "admin": False, "data": "public dashboard only"})


# =============================================================================
# SAFE REFERENCE — server-side ownership/visibility enforcement.
# The current user is simulated with ?current= (defaulting to 'alice'); a note
# is only returned when it is public OR owned by that user. Shown so the
# vulnerable/safe pair (this vs PERMUTATION 1) can be diffed by SAST and learners.
# =============================================================================
@bp.get("/note-safe")
def note_safe():
    note_id = request.args.get("id", "")
    current = request.args.get("current", "alice")
    db = get_db()
    row = db.execute(
        "SELECT id, owner, title, body, private FROM notes WHERE id = ?",
        (note_id,),
    ).fetchone()
    if row is None:
        return jsonify({"error": "not found", "id": note_id}), 404
    # SAFE: enforce visibility/ownership on the server before returning anything.
    if row["private"] and row["owner"] != current:
        return jsonify({"error": "forbidden", "id": note_id, "current": current}), 403
    return jsonify({"current": current, "note": dict(row)})

File: javascript/routes/access_control.js
/**
 * Broken Access Control — OWASP 2021 A01 | OWASP 2025 A01.
 *
 * Same annotation format as the SQL-injection reference module: each
 * permutation carries a standard header describing the flaw, its CWE, an
 * exploitation-likelihood rating, prevalence today, an example request, and
 * the corresponding safe pattern. Access-control flaws are business-logic
 * bugs — no framework enforces object- or function-level authorization for
 * you — so the vulnerable handlers below simply OMIT the check that should be
 * there. They are left unguarded on purpose so the demos are really
 * exploitable; a single clearly-labelled SAFE handler shows the fix.
 */
'use strict';

const express = require('express');
const { getDb } = require('../db');

const router = express.Router();

// ===========================================================================
// PERMUTATION 1 — IDOR: direct object reference with no ownership check
// OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
// CWE-639: Authorization Bypass Through User-Controlled Key
// EXPLOITATION LIKELIHOOD: HIGH — post-auth but trivial: the note id is a
//   sequential integer in the URL, so an attacker simply enumerates ?id=1,2,3…
//   and reads every user's PRIVATE notes (ids 2 and 4). No tooling, fully
//   deterministic, high-impact data exposure.
// PREVALENCE TODAY: greenfield MEDIUM (no framework/ORM enforces object-level
//   authorization — it is app-specific logic that is easy to forget; row-level
//   security and policy layers help only when teams actually add them) |
//   legacy/10yr tech-debt HIGH (CRUD-by-primary-key endpoints written before
//   ownership checks were routine remain everywhere; IDOR is the #1 A01 finding).
// EXAMPLE:
//   GET /access/note?id=2   -> alice's private "Bank PIN reminder"
//   GET /access/note?id=4   -> admin's private "Ops runbook" (prod db password)
// FIX: scope every fetch to the caller — WHERE id = ? AND (private = 0 OR
//   owner = ?) — using the SERVER-side session identity (see /access/note-safe).
// ===========================================================================
router.get('/note', (req, res) => {
  const id = req.query.id;
  // VULNERABLE: note fetched by id alone — no owner/private authorization check.
  const sql = 'SELECT id, owner, title, body, private FROM notes WHERE id = ?';
  try {
    const note = getDb().prepare(sql).get(id);
    res.json({ query: sql, id, note: note || null });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 2 — Missing function-level authorization on an admin endpoint
// OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
// CWE-862: Missing Authorization
// EXPLOITATION LIKELIHOOD: HIGH — the "admin" path is protected by nothing but
//   its name (security by URL obscurity). Any anonymous caller hits it directly
//   and dumps every user's password hash, email and SSN. Pre-auth, one request.
// PREVALENCE TODAY: greenfield MEDIUM (frameworks offer route guards/decorators,
//   but applying them is opt-in and a newly added admin/internal route is easily
//   shipped without one — deny-by-default is still not the norm) | legacy/10yr
//   tech-debt HIGH (admin/debug/export endpoints that assume "no link in the UI"
//   equals "unreachable" are a classic finding in older codebases).
// EXAMPLE:
//   GET /access/admin/users   -> full users table incl. md5 hashes + SSNs
// FIX: enforce authentication AND role at the boundary (require an authenticated
//   session, assert role === 'admin' via server-side state) before the handler
//   runs; deny by default.
// ===========================================================================
router.get('/admin/users', (req, res) => {
  // VULNERABLE: no authentication and no role check — anyone reaches this.
  const sql = 'SELECT id, username, password, email, role, ssn FROM users';
  try {
    const users = getDb().prepare(sql).all();
    res.json({ query: sql, note: 'no auth/role check performed', users });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 3 — Mass assignment / privilege escalation
// OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
// CWE-915: Improperly Controlled Modification of Dynamically-Determined Object
//          Attributes (mass assignment / auto-binding)
// EXPLOITATION LIKELIHOOD: HIGH — the profile update binds EVERY field present
//   in the JSON body straight onto the row, so a normal user adds one extra key
//   ("role":"admin") and escalates to administrator in a single request.
//   Deterministic, no tooling.
// PREVALENCE TODAY: greenfield MEDIUM (safe-by-default guards exist — DTOs,
//   zod/class-validator strict schemas, allow-listed columns — but the ORM
//   convenience patterns  Model.update(req.body)  and  {...user, ...req.body}
//   remain extremely common and re-open it whenever the schema step is skipped)
//   | legacy/10yr tech-debt HIGH (hand-rolled "update all posted fields" loops
//   over request bodies are a staple of older CRUD code).
// EXAMPLE:
//   curl -X POST /access/profile -H 'Content-Type: application/json' \
//        -d '{"username":"alice","email":"a@x.com","role":"admin"}'
//   -> alice.role becomes 'admin' even though role was never meant to be editable
// FIX: bind only an explicit allow-list of user-editable fields (never role);
//   e.g. destructure { email } = req.body and update just that column.
// ===========================================================================
router.post('/profile', (req, res) => {
  const body = req.body || {};
  const username = body.username;
  if (!username) {
    return res.status(400).json({ error: 'username required' });
  }
  // VULNERABLE: every posted field is written to the row — including role —
  // with no allow-list, so the client controls which columns get updated.
  const fields = Object.keys(body).filter((k) => k !== 'username');
  const setClause = fields.map((k) => `${k} = ?`).join(', ');
  const values = fields.map((k) => body[k]);
  const sql = `UPDATE users SET ${setClause} WHERE username = ?`;
  try {
    getDb().prepare(sql).run(...values, username);
    const updated = getDb()
      .prepare('SELECT id, username, email, role, ssn FROM users WHERE username = ?')
      .get(username);
    res.json({ query: sql, boundFields: fields, updated: updated || null });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 4 — Client-controlled authorization ("trust the X-Role header")
// OWASP 2021 A01 Broken Access Control  ->  OWASP 2025 A01 Broken Access Control
// CWE-602: Client-Side Enforcement of Server-Side Security
// EXPLOITATION LIKELIHOOD: HIGH — the admin decision is read from a value the
//   client fully controls (an "X-Role" request header or ?role= param). The
//   attacker just sets it to "admin" and the server hands back admin-only data.
//   Pre-auth, one request, no tooling.
// PREVALENCE TODAY: greenfield LOW (identity in new apps is derived from a
//   validated session/JWT verified server-side; trusting a raw request header
//   for authz is a textbook antipattern SAST/review flag — though it resurfaces
//   in service meshes that blindly trust gateway-set identity headers on a
//   directly-reachable service) | legacy/10yr tech-debt HIGH (hand-rolled auth
//   that reads a role from a header, cookie or hidden field and believes it is
//   common in older enterprise and internal apps).
// EXAMPLE:
//   curl -H 'X-Role: admin' /access/dashboard
//   GET  /access/dashboard?role=admin
//   -> returns admin-only data to an unauthenticated caller
// FIX: never derive privilege from client input; look up the role from the
//   server-side session/token for the authenticated principal.
// ===========================================================================
router.get('/dashboard', (req, res) => {
  // VULNERABLE: authorization decided from a client-supplied header/param.
  const role = req.get('X-Role') || req.query.role || 'user';
  const isAdmin = role === 'admin';
  if (isAdmin) {
    const secrets = getDb()
      .prepare('SELECT id, username, email, role, ssn FROM users')
      .all();
    return res.json({
      claimedRole: role,
      access: 'admin',
      adminData: { flag: 'ADMIN_DASHBOARD_UNLOCKED', users: secrets },
    });
  }
  res.json({ claimedRole: role, access: 'user', adminData: null });
});

// ===========================================================================
// SAFE REFERENCE — server-side ownership/role enforcement.
// The caller's identity comes from server-side state (here simulated by
// ?current=, defaulting to 'alice'; in a real app it is the authenticated
// session/token — never a note field or a client header). A note is returned
// only when it is public OR owned by the current user; a private note owned by
// someone else yields 403, so id enumeration leaks nothing.
// ===========================================================================
router.get('/note-safe', (req, res) => {
  const id = req.query.id;
  const current = req.query.current || 'alice'; // SAFE: identity from server side
  const note = getDb()
    .prepare('SELECT id, owner, title, body, private FROM notes WHERE id = ?')
    .get(id);
  if (!note) {
    return res.status(404).json({ id, note: null });
  }
  if (note.private && note.owner !== current) {
    return res.status(403).json({ id, error: 'forbidden', currentUser: current });
  }
  res.json({ id, currentUser: current, note });
});

module.exports = { base: '/access', router };
